### PR TITLE
feat: actually validate uuids

### PR DESCRIPTION
**Pull request by nordfjord**
So far the "parse" of the uuid modules has been kind of a lie, it just lowercases whatever you give it and pretends. This change will actually ensure that the Uuid is a hexadecimal string with 0 or 4 dashes.

Curious for opinions from @sroze and @everzet whether this is a good idea

### CODE REVIEW

**Comment by sroze** [comment]
Just so I understand what's the consequence on clients using the library? Does it change anything about the API contract or is this just an internal matter?

File: packages/core/src/lib/Uuid.ts [hunk 1]
```diff
@@ -2,6 +2,17 @@ import { randomUUID } from "crypto"
 
 export type Uuid<T> = string & { __brand: T }
 
+// permissive uuid regex that allows non-standard uuids
+const regex = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/
+
+function parse<T>(str: string): Uuid<T> {
+  const lower = str.toLowerCase()
+  if (!regex.test(str)) {
```
Comment (bartelink) [inline packages/core/src/lib/Uuid.ts:10]: Did you intend to test `lower`, or should L9 be inlined?
Could this create function become a gate that canonicalises to 'N` format (i.e. no dashes/braces etc, lowercase) ?

File: packages/core/src/lib/Uuid.ts [hunk 2]
```diff
@@ -2,6 +2,17 @@ import { randomUUID } from "crypto"
 
 export type Uuid<T> = string & { __brand: T }
 
+// permissive uuid regex that allows non-standard uuids
```
Comment (bartelink) [inline packages/core/src/lib/Uuid.ts:5]: doesnt look very permissive from where I'm sitting?

Comment (nordfjord) [reply packages/core/src/lib/Uuid.ts:5]: Permissive in that it accepts any hexadecimal string as long as it looks like a uuid. The canonical regex looks like this https://github.com/uuidjs/uuid/blob/main/src/regex.js

File: packages/core/src/lib/Uuid.ts [hunk 3]
```diff
@@ -2,6 +2,17 @@ import { randomUUID } from "crypto"
...
 
+// Custom regex: admits any hex values matching the normal dashed cluster grouping structure
+// loose in that it does not attempt to reject 'impossible/undefined' based on the UUID spec
+const regex = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i
+
+function parse<T>(str: string): Uuid<T> {
+  if (!regex.test(str)) {
+    throw new Error(`Uuid: invalid format '${str}'`)
+  }
+  return str.toLowerCase() as Uuid<T>
```
Comment (bartelink) [inline packages/core/src/lib/Uuid.ts:13]: if you're going to canonicalize as lower case, better to do that immediately and validate that the thing you'll ultimately store is valid/compliant
(and, to be pedantic/emphasize that, I'd make the regeex not be case-insensitive)

Comment (nordfjord) [reply packages/core/src/lib/Uuid.ts:13]: I don't really have a preference, so happy to go with that.

**Pull request merged**